app/queries/display_ads/filtered_ads_query.rb

```diff
@@ -72,6 +76,15 @@ def authenticated_ads(display_auth_audience)
       @filtered_display_ads.where(display_to: display_auth_audience)
     end
 
+    def user_targeting_ads
+      if @user_id
+        segment_ids = SegmentedUser.where(user_id: @user_id).pluck(:audience_segment_id)
+        @filtered_display_ads.where("audience_segment_id IS NULL OR audience_segment_id IN (?)", segment_ids)
+      else
+        @filtered_display_ads.where(audience_segment_id: nil)
```

Comment:
I was definitely feeling this out as I went. At one point, I didn't want to have that validation, because I had the idea that there might be some ads that would want an "or" option — like, these users _or_ logged-out visitors ("hey, turn dark mode on"), but that really does put some pressure on the ad creator to make sure it's targeting properly (we probably wouldn't want to have an ad targeting `trusted users` be anything other than `display_to: :logged_in`). 

But I was feeling the opposite when I got to writing this bit of logic — "better make sure we don't show a `trusted user` ad to logged-out folks". There are a couple of categories where "or" might make sense (`no posts yet`, `no experience level`) but many more where it wouldn't make sense at all. Maybe that's a signal that having an "or" here is a bit of a sharp edge?

If we lean on `display_to`, validate that we only set `audience_segment` if there's a `display_to: :logged_in`, then the query logic can be a bit cleaner.

So, I think it's a good question — I think _something_ here is a bit _off_ — but I don't have a strong feeling on what's our best option here.

Reply:
Yeah, makes sense to ensure we don't show segmented messages to non-segmented situations. What we have above is effectively a "soft validation" in that a segmented unit only ever shows up for `display_to: logged_in` no matter what.

Although we've currently sort of established that this is how it works. Instead of disallowing certain fields, we ignore them (i.e. when I set tags on an area which isn't going to care about tags — which is only policed via the UI, not validation).

So with that being the approach we have, I guess it's basically cool to just keep rolling with that — some configs will not show in certain situations, but we don't explicitly invalidate them.

With that being said, I think we could roll with the above, but keep an eye on how the logic is shaking out and potentially re-visiting this later with some of the other areas which are treating it the same way.